Python/executor.c

```diff
@@ -30,6 +30,14 @@
 #undef ENABLE_SPECIALIZATION
 #define ENABLE_SPECIALIZATION 0
 
+#undef SAVE_FRAME_STATE
+#define SAVE_FRAME_STATE() \
```

Comment:
Rather than a macro, I think the code generator needs to understand this.

Given that `SAVE_FRAME_STATE` is basically `SAVE_IP` followed by `_PyFrame_SetStackPointer(frame, stack_pointer);` we could convert it to two micro-ops: `SAVE_IP` and `SAVE_SP`.

In general, we want to avoid macros in the generated C code.
The generated code can be explicit and verbose.

Reply:
Well, there are already many macros (and static inline functions) in the generated code. The generator *recognizes* the presence of `SAVE_FRAME_STATE()`, but it doesn't *expand it* -- the C preprocessor can do that for us more easily. Currently we only do the expansion in the generator for things whose expansion requires information that only the generator has (like the stack adjustment for `ERROR_IF`).

Reply:
You are proposing that the macro expansion for `CALL_PY_EXACT_ARGS` become
```
        macro(CALL_PY_EXACT_ARGS) =
            unused/1 + // Skip over the counter
            _CHECK_PEP_523 +
            _CHECK_FUNCTION_EXACT_ARGS +
            _CHECK_STACK_SPACE +
            _INIT_CALL_PY_EXACT_ARGS +
            SAVE_IP +  // Tier 2 only; special-cased oparg
            SAVE_CURRENT_IP +  // <------------------- Addition
            _PUSH_FRAME;
```
where `SAVE_CURRENT_IP` is something like this:
```
op(SAVE_CURRENT_IP, (--)) {
    #if TIER_ONE
    frame->prev_instr = next_instr - 1;
    #endif
    #if TIER_TWO
    frame->prev_instr--;
    #endif
}
```
Or we could special-case its expansion in the generator, potayto-potato. But it *has* to differ between tiers because in Tier 1 it must store next_instr whereas in Tier 2 it must rely on the preceding SAVE_IP to set `frame->prev_instr`. (Ideally at some point in the future we won't need the `prev_instr--` yet, but that's a tricky change.)

The `_PyFrame_SetStackPointer(frame, stack_pointer);` call should be moved back into `_PUSH_FRAME` (at the point where I currently call `SAVE_FRAME_STATE`).

If I can get this to work I'll apply it and merge the PR.

Reply:
I *did* get this working (see https://github.com/python/cpython/pull/107760/commits/05af84805a04747c1de7a5b843041b8d55c31bee), and will test and benchmark it before merging it.

Note that there are still some `#if TIER_ONE` and `#if TIER_TWO` sections, but they are unavoidable.